src/jwt-utils/jwtdecode.ts

```diff
@@ -0,0 +1,192 @@
+import crypto from 'crypto'
+
+import { PublicKey } from '..'
+import * as base64UrlSafe from '../base64urlsafe'
+import { JwtVerifyError } from '../jwtverifyerror'
+
+const defaultOptions = {
+  expiresSkew: 0,
+  expiresMax: 0,
+  nbfIatSkew: 300,
+  fixup: null
+}
+
+export function decode(
+  jwt: string,
+  publicKeys: Record<string, Record<string, string | PublicKey>>,
+  audiences: string[],
+  options: Record<string, unknown> | number = defaultOptions
+): Record<string, string | number> {
+  if (typeof options === 'number') {
+    // Backwards compatibility with old api
+    options = {
+      ...defaultOptions,
+      nbfIatSkew: options
+    }
+  }
+
+  if (typeof options !== 'object' || Array.isArray(publicKeys)) {
+    throw new Error('options needs to a map of { nbfIatSkew: 300, ... }')
+  }
+
+  const parts = jwt.split(/\./)
+  if (parts.length !== 3) {
+    throw new JwtVerifyError('JWT does not contain 3 dots')
+  }
+
+  const header = JSON.parse(base64UrlSafe.decode(parts[0]).toString('utf8'))
+  const body = JSON.parse(base64UrlSafe.decode(parts[1]).toString('utf8'))
+  if (typeof options.fixup === 'function') {
+    options.fixup(header, body)
+  }
+
+  let signAlgo = null
+  let hmacAlgo = null
+  switch (header.alg) {
+    case 'RS256':
+      signAlgo = 'RSA-SHA256'
+      break
+    case 'RS384':
+      signAlgo = 'RSA-SHA384'
+      break
+    case 'RS512':
+      signAlgo = 'RSA-SHA512'
+      break
+    case 'ES256':
+      signAlgo = 'sha256'
+      break
+    case 'ES384':
+      signAlgo = 'sha384'
+      break
+    case 'ES512':
+      signAlgo = 'sha512'
+      break
+    case 'HS256':
+      hmacAlgo = 'sha256'
+      break
+    case 'HS384':
+      hmacAlgo = 'sha384'
+      break
+    case 'HS512':
+      hmacAlgo = 'sha512'
+      break
+    default:
+      throw new JwtVerifyError(
+        'Only alg RS256, RS384, RS512, ES256, ES384, ES512, HS256, HS384 and HS512 are supported'
+      )
+  }
+
+  if (!body.iss) {
+    throw new JwtVerifyError('No issuer set')
+  }
+
+  const issuer = publicKeys[body.iss]
+  if (!issuer) {
+    throw new JwtVerifyError(`Unknown issuer '${body.iss}'`)
+  }
+
+  // Find public key
+  let pubkeyOrSharedKey =
+    typeof header.kid === 'string'
+      ? (issuer[`${header.kid}@${header.alg}`] as string)
+      : (issuer[`default@${header.alg}`] as string)
+
+  let issuerOptions: Record<string, unknown> = {}
+  if (typeof pubkeyOrSharedKey === 'object' && pubkeyOrSharedKey !== null && pubkeyOrSharedKey['publicKey']) {
+    issuerOptions = pubkeyOrSharedKey
+    pubkeyOrSharedKey = pubkeyOrSharedKey['publicKey']
+  }
+
+  if (!pubkeyOrSharedKey) {
+    throw new JwtVerifyError(`Unknown pubkey id '${header.kid}' for this issuer`)
+  }
+
+  const signatureOrHash = base64UrlSafe.decode(parts[2])
+  /* istanbul ignore else */
```

Comment:
I don't believe this is necessary any longer

Reply:
I'm not sure what you mean by this. The `signatureOrHash` value is used just below to verify signatures for certain algorithms if I'm understanding correctly. Do we not need to verify signatures anymore? Or is there another way to do that?